File: src/branchnexus/hooks/runner.py
"""Panel command hooks runner."""

from __future__ import annotations

import logging as py_logging
import shlex
import subprocess
from dataclasses import dataclass

logger = py_logging.getLogger(__name__)


@dataclass(frozen=True)
class HookExecution:
    command: str
    success: bool
    returncode: int
    output: str


@dataclass
class HookRunResult:
    pane: int
    executions: list[HookExecution]

    @property
    def has_failures(self) -> bool:
        return any(not execution.success for execution in self.executions)


class HookRunner:
    def __init__(
        self,
        timeout_seconds: float = 30.0,
        *,
        trusted_config: bool = True,
        allow_command_prefixes: list[str] | None = None,
    ) -> None:
        self.timeout_seconds = timeout_seconds
        self.trusted_config = trusted_config
        self.allow_command_prefixes = [item.strip() for item in (allow_command_prefixes or []) if item.strip()]

    def _is_command_allowed(self, command: str) -> bool:
        if self.trusted_config:
            return True
        try:
            argv = shlex.split(command)
        except ValueError:
            return False
        if not argv:
            return False
        if not self.allow_command_prefixes:
            return False
        return argv[0] in set(self.allow_command_prefixes)

    def run(
        self,
        *,
        pane: int,
        commands: list[str],
        runner: callable = subprocess.run,
    ) -> HookRunResult:
        executions: list[HookExecution] = []
        logger.debug("Running %s hook commands for pane=%s", len(commands), pane)
        for command in commands:
            if not self._is_command_allowed(command):
                logger.warning("Hook command blocked by policy pane=%s command=%s", pane, command)
                executions.append(
                    HookExecution(
                        command=command,
                        success=False,
                        returncode=126,
                        output="Command blocked by hook trust policy.",
                    )
                )
                continue
            try:
                logger.debug("Executing hook command pane=%s command=%s", pane, command)
                argv = ["bash", "-lc", command]
                completed = runner(
                    argv,
                    shell=False,
                    capture_output=True,
                    text=True,
                    timeout=self.timeout_seconds,
                    check=False,
                )
                success = completed.returncode == 0
                output = f"{completed.stdout}{completed.stderr}".strip()
                if not success:
                    logger.warning(
                        "Hook command failed pane=%s returncode=%s command=%s",
                        pane,
                        completed.returncode,
                        command,
                    )
                executions.append(
                    HookExecution(
                        command=command,
                        success=success,
                        returncode=completed.returncode,
                        output=output,
                    )
                )
            except subprocess.TimeoutExpired:
                logger.error("Hook command timed out pane=%s command=%s", pane, command)
                executions.append(
                    HookExecution(
                        command=command,
                        success=False,
                        returncode=124,
                        output="Command timed out.",
                    )
                )
        return HookRunResult(pane=pane, executions=executions)

File: ts-src/hooks/runner.ts
import { execa } from 'execa';
import { logger } from '../utils/logger.js';
import { buildWslCommand } from '../runtime/wsl.js';
import { Platform, detectPlatform } from '../runtime/platform.js';
import { hasDistribution } from '../utils/validators.js';

export interface HookExecution {
  command: string;
  success: boolean;
  returncode: number;
  output: string;
}

export interface HookRunResult {
  pane: number;
  executions: HookExecution[];
  hasFailures: boolean;
}

export class HookRunner {
  private timeoutSeconds: number;
  private trustedConfig: boolean;
  private allowCommandPrefixes: string[];

  constructor(options?: {
    timeoutSeconds?: number;
    trustedConfig?: boolean;
    allowCommandPrefixes?: string[];
  }) {
    this.timeoutSeconds = options?.timeoutSeconds ?? 30;
    this.trustedConfig = options?.trustedConfig ?? true;
    this.allowCommandPrefixes = options?.allowCommandPrefixes ?? [];
  }

  private isCommandAllowed(command: string): boolean {
    if (this.trustedConfig) {
      return true;
    }

    const argv = command.split(/\s+/);
    if (argv.length === 0) {
      return false;
    }

    if (this.allowCommandPrefixes.length === 0) {
      return false;
    }

    return this.allowCommandPrefixes.includes(argv[0]);
  }

  async run(pane: number, commands: string[], distribution?: string): Promise<HookRunResult> {
    const executions: HookExecution[] = [];
    logger.debug(`Running ${commands.length} hook commands for pane=${pane}`);

    for (const command of commands) {
      if (!this.isCommandAllowed(command)) {
        logger.warn(`Hook command blocked by policy pane=${pane} command=${command}`);
        executions.push({
          command,
          success: false,
          returncode: 126,
          output: 'Command blocked by hook trust policy.',
        });
        continue;
      }

      try {
        logger.debug(`Executing hook command pane=${pane} command=${command}`);

        const isWindows = detectPlatform() === Platform.WINDOWS;
        const cmd = ['bash', '-lc', command];
        const finalCmd =
          isWindows && hasDistribution(distribution) ? buildWslCommand(distribution, cmd) : cmd;

        const result = await execa(finalCmd[0], finalCmd.slice(1), {
          timeout: this.timeoutSeconds * 1000,
          reject: false,
          all: true,
        });

        const success = result.exitCode === 0;
        const output = `${result.stdout}${result.stderr}`.trim();

        if (!success) {
          logger.warn(
            `Hook command failed pane=${pane} returncode=${result.exitCode} command=${command}`
          );
        }

        executions.push({
          command,
          success,
          returncode: result.exitCode,
          output,
        });
      } catch (error) {
        const isTimeout = error instanceof Error && error.message.includes('timed out');
        logger.error(
          `Hook command ${isTimeout ? 'timed out' : 'failed'} pane=${pane} command=${command}`
        );

        executions.push({
          command,
          success: false,
          returncode: isTimeout ? 124 : 1,
          output: isTimeout
            ? 'Command timed out.'
            : error instanceof Error
              ? error.message
              : 'Unknown error',
        });
      }
    }

    return {
      pane,
      executions,
      hasFailures: executions.some((e) => !e.success),
    };
  }
}
